packages/application/src/index.ts

```diff
@@ -186,31 +210,31 @@ export class Application<T extends Widget> {
    * @param plugin - The plugin to register.
    *
    * #### Notes
-   * An error will be thrown if a plugin with the same id is already
+   * An error will be thrown if a plugin with the same ID is already
    * registered, or if the plugin has a circular dependency.
    *
    * If the plugin provides a service which has already been provided
    * by another plugin, the new service will override the old service.
    */
   registerPlugin(plugin: IPlugin<this, any>): void {
-    // Throw an error if the plugin id is already registered.
-    if (plugin.id in this._pluginMap) {
-      throw new Error(`Plugin '${plugin.id}' is already registered.`);
+    // Throw an error if the plugin ID is already registered.
+    if (this._plugins.has(plugin.id)) {
+      throw new TypeError(`Plugin '${plugin.id}' is already registered.`);
```

Comment:
"The TypeError object represents an error when an operation could not be performed, typically (but not exclusively) when a value is not of the expected type."

This doesn't really seem appropriate here?

Reply:
I just wondered why a `TypeError` was chosen for all the errors thrown in this PR. The only argument for it would seem to be that it adds some context clues, and if the context isn't that you passed something of the wrong type, then it is only confusing.

Reply:
There are three `ReferenceError`s, two `TypeError`s, one *new* `Error` and a couple old ones.

Reply:
It was an attempt to make the errors more semantic, but if you thing it detracts, there's no technical need for them to be any specific variety of error.

Reply:
It was mostly that my impression was that `TypeError` and `ReferenceError` has certain *specific* meanings (the type of something was wrong, typically via `instanceof` check; and an unassigned variable was attempted to be dereferenced).

Reply:
I was checking the MDN pages for the errors when I picked these errors.

> The [**ReferenceError**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ReferenceError) object represents an error when a variable that doesn't exist (or hasn't yet been initialized) in the current scope is referenced.

My rationale was that not being registered yet is akin to "hasn't yet been initialized".

> A [**TypeError**](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError) may be thrown when:
> 
>    * an operand or argument passed to a function is incompatible with the type expected by that operator or function; or
>    * when attempting to modify a value that cannot be changed; or
>    * when attempting to use a value in an inappropriate way.

I think, for example, if a plugin does not have a `deactivate` method, trying to deactivate it is using it in an inappropriate way.

---

I don't feel dogmatic about these choices. If you feel they are inappropriate, I'm happy to change them back to generic `Error` instances. But I do think a reasonable reading of the documentation for these errors allows some wiggle room as well.

Do you think we should change it?